crates/sui-core/src/transaction_manager.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
+
+use sui_storage::node_sync_store::NodeSyncStore;
+use sui_types::{
+    base_types::TransactionDigest, committee::EpochId, error::SuiResult,
+    messages::VerifiedCertificate,
+};
+use tokio::sync::mpsc::UnboundedSender;
+use tracing::error;
+
+use crate::authority::{authority_store::ObjectKey, AuthorityMetrics, AuthorityStore};
+
+/// TransactionManager is responsible for managing pending certificates and publishes a stream
+/// of certificates ready to be executed. It works together with AuthorityState for receiving
+/// pending certificates, and getting notified about committed objects. Executing driver
+/// subscribes to the stream of ready certificates published by the TransactionManager, and can
+/// execute them in parallel.
+/// TODO: use TransactionManager for fullnode.
+pub(crate) struct TransactionManager {
```

Comment:
A quick sketch of a design for handling unsequenced certs:

1. pending certs can be in one of three states: Unsequenced, InputsMissing, and Executable.
2. Certs start in the Unsequenced state.
3. If they have no shared objects, they advance to InputsMissing.
4. If they have shared objects, we call `is_certificate_awaiting_sequencing()` to determine whether they advance to InputMissing.
5. The consensus task notifies TransactionManager when a cert has been sequenced, so that it can advance to InputsMissing.
6. The InputsMissing->Executable transition is of course the piece that you've already implemented.

Reply:
This should be implementable, but I don't see the benefit yet and it does add complexity. Can we keep the TransactionManager simple for now, until we need to support more use cases e.g. for checkpoint sync output?

Reply:
Added a todo to investigate taking shared object locks inside TransactionManager.